### DISCUSSION TITLE
Splink Errors on EMR

### DISCUSSION THREAD

**Discussion by pranav2902** [moj-analytical-services/splink#1628] [Q&A]
Hello Team,

I am working with Splink on Spark backend to solve the record linkage problem at scale. I am running PySpark jobs on EMR Serverless. For linking, my target left sided dataset is of size 5 million records, and the right sided one is of 50 million records. I have chunked the right sided datasets into smaller chunks, of around 1-5 million records each, for simpler testing. The spark parameters i am using are: Enabling arrow execution, and setting the parallelism to around 5*number of CPUs in the cluster, as suggested by the docs.

EMR Serverless application specs - 50 workers, with each worker having 4 vCPUs, 24 GB memory and 200 GB Disk space. spark.sql.parallelism is hence set to 1000

Splink version : 3.9.3
Spark version : 3.4.1

I am running into various errors on trying to run the Splink record linkage job on EMR serverless. I have 4 blocking rules, and three of those together work perfectly fine. The fourth one is a blocking rule which goes as:
"split(l.name,' ')[0] = split(r.name,' ')[0]", where name is a normalized field, and i am checking for a first token match in the name. This blocking rule generates far more comparisons as compared to the other three.

Including this fourth rule causes a variety of issues. If i use salting, i get the following error in the linker.predict step:

```
  File "/home/hadoop/environment/lib64/python3.7/site-packages/sqlglot/parser.py", line 2567, in _parse_type
    this = self._parse_column()
  File "/home/hadoop/environment/lib64/python3.7/site-packages/sqlglot/parser.py", line 2711, in _parse_column
    this = self._parse_field()
  File "/home/hadoop/environment/lib64/python3.7/site-packages/sqlglot/parser.py", line 2807, in _parse_field
    return self._parse_primary() or self._parse_function() or self._parse_id_var(any_token)
  File "/home/hadoop/environment/lib64/python3.7/site-packages/sqlglot/parser.py", line 3524, in _parse_id_var
    return exp.Identifier(this=prefix + self._prev.text, quoted=quoted)
  File "/home/hadoop/environment/lib64/python3.7/site-packages/sqlglot/expressions.py", line 98, in __init__
    self._set_parent(arg_key, value)
  File "/home/hadoop/environment/lib64/python3.7/site-packages/sqlglot/expressions.py", line 280, in _set_parent
    if isinstance(value, Expression):
RecursionError: maximum recursion depth exceeded while calling a Python object
```

If I dont use salting, I get the following error, again in the linker.predict step - No space left on device:

```
py4j.protocol.Py4JJavaError: An error occurred while calling o837.checkpoint.
: org.apache.spark.SparkException: Job aborted due to stage failure: Task 11 in stage 373.0 failed 4 times, most recent failure: Lost task 11.3 in stage 373.0 (TID 19747) ([2600:1f13:ac0:8500:8912:9981:2295:3902] executor 111): 
java.io.IOException: No space left on device
```

I am using salting by specifying my settings dictionary, with blocking rules given as a list of dictionaries, with each having the "blocking_rule" and "salting_partitions" keys. If Rule A generates 10 comparisons and Rule B generates 2 comparisons, I loosely give the salting_partitions in the ratio 5:1 for A:B. Please correct me if I am doing salting incorrectly!

I am unable to solve both of these issues, it would be great to get your help/inputs to resolve these!

**Comment by RobinL** [comment]
You're doing thing correctly - it's probably just that your blocking rule generates too many comparisons to be feasibly computable.  Have you tried using  [count_num_comparisons_from_blocking_rule](https://moj-analytical-services.github.io/splink/linkerbloc.html?h=count_#splink.linker.Linker.count_num_comparisons_from_blocking_rule) to see how many comparisons would be generated by your rule?

Another thing to check is the behaviour of `split(l.name,' ')[0]` if you have any data quality issues.
- what does that return if the name is blank
- what does it return if the name begins with a space

Is it possible that function is return a large number of values which are exactly the same?  If so, supposing it return 1,000,000 such records, that would be responsible for about 500m comparisons!
